### PR TITLE
[Capture] Add finite differences jvps

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

Currently, program capture only supports differentiation with backprop. This is only supported on `default.qubit`

**Description of the Change:**

Adds the capability to take finite difference derivatives.

**Benefits:**

We can now differentiate anything, and we can differentiate with the lightnings.  

**Possible Drawbacks:**

finite diff tends to be noisy to the point of uselessness.

**Related GitHub Issues:**

[sc-82167]

### CODE REVIEW

File: pennylane/workflow/_capture_qnode.py [hunk 1]
```diff
@@ -291,7 +290,50 @@ def _backprop(args, tangents, **impl_kwargs):
...
-diff_method_map = {"backprop": _backprop}
+def _finite_diff(args, tangents, **impl_kwargs):
+
+    gradient_kwargs = impl_kwargs["qnode_kwargs"]["gradient_kwargs"]
+    h = gradient_kwargs.get("h", 1e-6)
+    if gradient_kwargs.get("approx_order", 1) != 1:
+        raise NotImplementedError("only approx_order=1 is currently supported.")
+    if gradient_kwargs.get("strategy", "forward") != "forward":
+        raise NotImplementedError("only strategy='forward' is currently supported.")
+    available_kwargs = {"h", "strategy", "approx_order"}
```
Comment (andrijapau) [inline pennylane/workflow/_capture_qnode.py:301]: This might be overkill but is it worth considering standardizing what gradient kwarg belongs to which gradient transform? Do we ever expect these names to change?
```suggestion
    # this could be stored in gradient_transform.py
    FINITE_DIFF_SUPPORTED_KWARGS = {"h", "strategy", "approx_order"}
    available_kwargs = FINITE_DIFF_SUPPORTED_KWARGS
```

File: tests/capture/workflow/test_capture_finite_diff.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    with pytest.warns(UserWarning, match="Detected float32 parameter with finite differences."):
+        jax.grad(circuit)(jnp.array(0.5, dtype=jnp.float32))
+
+
+class TestGradients:
+
+    @pytest.mark.parametrize("grad_f", (jax.grad, jax.jacobian))
+    def test_simple_circuit(self, grad_f):
+        """Test accurage results for a simple, single parameter circuit."""
```
Comment (mudit2812) [inline tests/capture/workflow/test_capture_finite_diff.py:45]:
```suggestion
        """Test accurate results for a simple, single parameter circuit."""
```

File: tests/capture/workflow/test_capture_finite_diff.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+def test_warning_float32():
+    """Test that a warning is raised if trainable inputs are float32."""
+
+    @qml.qnode(qml.device("default.qubit", wires=1), diff_method="finite-diff")
+    def circuit(x):
+        qml.RX(x, 0)
+        return qml.expval(qml.Z(0))
+
+    with pytest.warns(UserWarning, match="Detected float32 parameter with finite differences."):
+        jax.grad(circuit)(jnp.array(0.5, dtype=jnp.float32))
```
Comment (mudit2812) [inline tests/capture/workflow/test_capture_finite_diff.py:38]: Should we be using `qml.grad` in the tests?

Comment (mudit2812) [reply tests/capture/workflow/test_capture_finite_diff.py:38]: Potential over-testing suggestion: or maybe parametrizing over qml and jax diff functions

File: tests/capture/workflow/test_capture_finite_diff.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+
+class TestGradients:
```
Comment (mudit2812) [inline tests/capture/workflow/test_capture_finite_diff.py:41]: Non blocking questions:
Do we have tests to validate that the correct diff method is actually being used? Does it matter if we do?

File: pennylane/gradients/finite_difference.py [hunk 5]
```diff
@@ -166,6 +167,88 @@ def finite_diff_coeffs(n, approx_order, strategy):
...
+    (1.2, 0.25)
+    >>> dresults
+    [np.float64(3.399999999986747), np.float64(1.000001000006634)]
+
+    """
+    coeffs, shifts = finite_diff_coeffs(n=1, approx_order=approx_order, strategy=strategy)
+
+    initial_res = f(*args)
+    if not isinstance(initial_res, (list, tuple)):
+        raise ValueError("Input function f must return either a list or tuple.")
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/gradients/finite_difference.py:217]: It's designed to match the jax syntax for a jvp rule.  But potentially I could just use pytrees.

File: tests/capture/workflow/test_capture_finite_diff.py (hunk 3 above)
Comment (albi3ro) [reply tests/capture/workflow/test_capture_finite_diff.py:38]: This is a jax specific test.

File: tests/capture/workflow/test_capture_finite_diff.py (hunk 4 above)
Comment (albi3ro) [reply tests/capture/workflow/test_capture_finite_diff.py:41]: I have a test at the bottom where I double check the captured jaxpr.

File: pennylane/gradients/finite_difference.py [hunk 6]
```diff
@@ -166,6 +167,92 @@ def finite_diff_coeffs(n, approx_order, strategy):
     return coeffs_and_shifts
 
 
+def finite_diff_jvp(
+    f: Callable,
+    args: tuple,
+    tangents: tuple,
+    *,
+    h: float = 1e-6,
```
Comment (andrijapau) [inline pennylane/gradients/finite_difference.py:175]: Why do we use `1e-6` here and `1e-7` elsewhere in the file?

File: tests/capture/workflow/test_capture_finite_diff.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    def test_hessian(self):
+        """Test that higher order derivatives like the hessian can be computed."""
+
+        @qml.qnode(qml.device("default.qubit", wires=4), diff_method="finite-diff")
+        def circuit(x):
+            qml.RX(x, 0)
+            return qml.expval(qml.Z(0))
+
+        hess = jax.grad(jax.grad(circuit))(0.5)
+        print(hess - -jnp.cos(0.5))
```
Comment (andrijapau) [inline tests/capture/workflow/test_capture_finite_diff.py:177]:
```suggestion
```

File: pennylane/gradients/finite_difference.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/gradients/finite_difference.py:175]: I think this was a byproduct of 1e-7 not meshing well with the default float32 of jax. But now there is a warning, so maybe we can decrease it. Just a bit wary since 1e-7 gives zero results with the default jax setup.

File: pennylane/gradients/finite_difference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/gradients/finite_difference.py:218]: Seems like a nice-to-have/backlog task. Mind creating a backlog story for it?

File: pennylane/gradients/finite_difference.py [hunk 8]
```diff
@@ -220,11 +221,14 @@ def finite_diff_jvp(
...
+        t = np.array(t) if isinstance(t, (int, float, complex)) else t
+
+
+        if qml.math.get_dtype_name(args[i]) in ("float32", "complex64"):
 
-        if qml.math.get_dtype_name(args[i]) == "float32":
             warn(
-                "Detected float32 parameter with finite differences. Recommend use of float64 with finite diff.",
+                "Detected 32 bits precision parameter with finite differences. Recommend use of 64 bits precision with finite diff.",
+
```
Comment (mudit2812) [inline pennylane/gradients/finite_difference.py:231]:
```suggestion
```

File: pennylane/gradients/finite_difference.py [hunk 9]
```diff
@@ -166,6 +167,92 @@ def finite_diff_coeffs(n, approx_order, strategy):
...
+    for i, t in enumerate(tangents):
+        if type(t).__name__ == "Zero":  # Zero = jax.interpreters.ad.Zero
+            continue
+        t = np.array(t) if isinstance(t, (int, float, complex)) else t
+
+
+        if qml.math.get_dtype_name(args[i]) in ("float32", "complex64"):
+
+            warn(
+                "Detected 32 bits precision parameter with finite differences. Recommend use of 64 bits precision with finite diff.",
```
Comment (andrijapau) [inline pennylane/gradients/finite_difference.py:230]:
```suggestion
                "Detected 32 bits precision parameter with finite differences. It is recommended to use  64 bit precision when computing finite differences.",
```

File: pennylane/gradients/finite_difference.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/gradients/finite_difference.py:175]: downgrading it back to 1e-7.

File: pennylane/gradients/finite_difference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/gradients/finite_difference.py:218]: eh... I just added it to this PR.

File: tests/capture/workflow/test_capture_finite_diff.py (hunk 7 above)
Comment (albi3ro) [reply tests/capture/workflow/test_capture_finite_diff.py:177]: resolved locally

File: pennylane/gradients/finite_difference.py [hunk 10]
```diff
@@ -166,6 +167,88 @@ def finite_diff_coeffs(n, approx_order, strategy):
...
+    jvps = [0 for _ in flat_initial_res]
+    for i, t in enumerate(tangents):
+        if type(t).__name__ == "Zero":  # Zero = jax.interpreters.ad.Zero
+            continue
+        t = np.array(t) if isinstance(t, (int, float, complex)) else t
+
+        if qml.math.get_dtype_name(args[i]) in ("float32", "complex64"):
+
+            warn(
+                "Detected 32 bits precision parameter with finite differences. It is recommended to use  64 bit precision when computing finite differences.",
```
Comment (andrijapau) [inline pennylane/gradients/finite_difference.py:227]: Extra space lol
```suggestion
                "Detected 32 bits precision parameter with finite differences. It is recommended to use 64 bit precision when computing finite differences.",

```

File: pennylane/gradients/finite_difference.py [hunk 11]
```diff
@@ -172,20 +172,20 @@ def finite_diff_jvp(
...
-    h: float = 1e-6,
+    h: float = 1e-7,
     approx_order: int = 1,
     strategy: Literal["forward", "backward", "center"] = "forward",
 ) -> tuple:
     r"""Compute the jvp of a generic function using finite differences.
 
     Args:
-        f (Callable): a generic function that returns an iterable of tensors. Note that this
+        f (Callable): a generic function that a pytree of tensors. Note that this
```
Comment (mudit2812) [inline pennylane/gradients/finite_difference.py:182]:
```suggestion
        f (Callable): a generic function that returns a pytree of tensors. Note that this

```

**Review by mudit2812 — APPROVED** [review]

File: pennylane/gradients/finite_difference.py [hunk 12]
```diff
@@ -166,6 +167,88 @@ def finite_diff_coeffs(n, approx_order, strategy):
...
+    approx_order: int = 1,
+    strategy: Literal["forward", "backward", "center"] = "forward",
+) -> tuple:
+    r"""Compute the jvp of a generic function using finite differences.
+
+    Args:
+        f (Callable): a generic function that a pytree of tensors. Note that this
+            function should not have keyword arguments.
+        args (tuple[TensorLike]): The tuple of arguments to ``f``
+        tangents (tuple[TensorLike]): the tuple of tangents for the arguments.
```
Comment (andrijapau) [inline pennylane/gradients/finite_difference.py:185]:
```suggestion
        tangents (tuple[TensorLike]): the tuple of tangents for the arguments ``args``

```

File: pennylane/gradients/finite_difference.py [hunk 13]
```diff
@@ -166,6 +167,88 @@ def finite_diff_coeffs(n, approx_order, strategy):
...
+            ``"forward"``, ``"center"``, or ``"backward"``.
+            For the ``"forward"`` strategy, the finite-difference shifts occur at the points
+            :math:`x_0, x_0+h, x_0+2h,\dots`, where :math:`h` is some small
+            stepsize. The ``"backwards"`` strategy is similar, but in
+            reverse: :math:`x_0, x_0-h, x_0-2h, \dots`. Finally, the
+            ``"center"`` strategy results in shifts symmetric around the
+            unshifted point: :math:`\dots, x_0-2h, x_0-h, x_0, x_0+h, x_0+2h,\dots`.
+
+    Returns:
+        tuple(TensorLike, TensorLike): The results and the cotangents of the results
```
Comment (andrijapau) [inline pennylane/gradients/finite_difference.py:200]:
```suggestion
        tuple(TensorLike, TensorLike): the results and their cotangents

```
